src/Framework/AspNetCoreAnalyzers/test/WebApplicationBuilder/PublicPartialProgramClassTest.cs

```diff
@@ -0,0 +1,217 @@
+    }
+
+    [Fact]
+    public async Task RemovesDeclarationIfItIsFound_WithLeadingTrivia()
+    {
+        // Arrange
+        var source = """
+using Microsoft.AspNetCore.Builder;
+
+var app = WebApplication.Create();
+
+app.MapGet("/", () => "Hello, World!");
+
+app.Run();
+
+// This is a test
+
+{|#0:public partial class Program;|}
+""";
+
+        var diagnostic = new DiagnosticResult(DiagnosticDescriptors.PublicPartialProgramClassNotRequired)
+                .WithLocation(0);
+
+        var fixedSource = """
+using Microsoft.AspNetCore.Builder;
+
+var app = WebApplication.Create();
+
+app.MapGet("/", () => "Hello, World!");
+
+app.Run();
+
+// This is a test
+
+
+""";
+
+        // Assert
+        await VerifyCS.VerifyCodeFixAsync(source, [diagnostic], fixedSource);
+    }
+
+    [Fact]
+    public async Task DoesNotGeneratesSource_IfProgramDeclaresExplicitInternalAccess()
+    {
+        var source = """
+using Microsoft.AspNetCore.Builder;
+
+var app = WebApplication.Create();
+
+app.MapGet("/", () => "Hello, World!");
+
+app.Run();
+
+internal partial class Program { }
+""";
+
+        await VerifyCS.VerifyCodeFixAsync(source, source);
+    }
+
+    [Fact]
+    public async Task DoesNotFix_ExplicitPublicProgramClass()
+    {
+        var source = """
+using Microsoft.AspNetCore.Builder;
+
+public class Program
+{
+    public static void Main()
+    {
+        var app = WebApplication.Create();
+
+        app.MapGet("/", () => "Hello, World!");
+
+        app.Run();
+    }
+}
+""";
+
+        await VerifyCS.VerifyCodeFixAsync(source, source);
+    }
+
+    [Fact]
+    public async Task DoesNotFix_ExplicitPublicPartialProgramClass()
+    {
+        var source = """
+using Microsoft.AspNetCore.Builder;
+
+public partial class Program
+{
+    public static void Main()
+    {
+        var app = WebApplication.Create();
+
+        app.MapGet("/", () => "Hello, World!");
+
+        app.Run();
+    }
+}
+""";
+
+        await VerifyCS.VerifyCodeFixAsync(source, source);
+    }
+
+    [Fact]
+    public async Task DoesNotFix_ExplicitInternalProgramClass()
+    {
+        var source = """
+using Microsoft.AspNetCore.Builder;
+
+internal class Program
+{
+    public static void Main()
+    {
+        var app = WebApplication.Create();
+
+        app.MapGet("/", () => "Hello, World!");
+
+        app.Run();
+    }
+}
+""";
+
+        await VerifyCS.VerifyCodeFixAsync(source, source);
+    }
+
+    [Theory]
+    [InlineData("interface")]
+    [InlineData("struct")]
+    public async Task DoesNotFix_ExplicitInternalProgramType(string type)
+    {
+        var source = $$"""
+using Microsoft.AspNetCore.Builder;
+
+internal {{type}} Program
+{
+    public static void Main(string[] args)
+    {
+        var app = WebApplication.Create();
+
+        app.MapGet("/", () => "Hello, World!");
+
+        app.Run();
+    }
+}
+""";
+
+        await VerifyCS.VerifyCodeFixAsync(source, source);
+    }
+}
```

Comment:
maybe a test for what happens when the offending `public partial class Program {}` is in a namespace or nested type? I don't expect either to be a problem, but worth proving it?

Reply:
Yep, I think our `CompilationUnitSyntax` check resolves this but I added a test as well.